rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/telemetry/TbMsgTimeseriesNodeTest.java

```diff
@@ -0,0 +1,242 @@
+        var tenantProfile = getTenantProfile();
+        when(ctxMock.getTenantProfile()).thenReturn(tenantProfile);
+        node.init(ctxMock, configuration);
+
+        String data = """
+                {
+                    "temp": 45,
+                    "humidity": 77
+                }
+                """;
+        long ts = System.currentTimeMillis();
+        var metadata = Map.of("ts", String.valueOf(ts));
+        TbMsg msg = TbMsg.newMsg(TbMsgType.POST_TELEMETRY_REQUEST, DEVICE_ID, new TbMsgMetaData(metadata), data);
+
+        when(ctxMock.getTelemetryService()).thenReturn(telemetryServiceMock);
+        when(ctxMock.getTenantId()).thenReturn(TENANT_ID);
+        doAnswer(invocation -> {
+            TelemetryNodeCallback callback = invocation.getArgument(5);
+            callback.onSuccess(null);
+            return null;
+        }).when(telemetryServiceMock).saveWithoutLatestAndNotify(any(), any(), any(), anyList(), anyLong(), any());
+
+        node.onMsg(ctxMock, msg);
+
+        verify(ctxMock).addTenantProfileListener(any());
+        ArgumentCaptor<List<TsKvEntry>> entryListCaptor = ArgumentCaptor.forClass(List.class);
+        verify(telemetryServiceMock).saveWithoutLatestAndNotify(
+                eq(TENANT_ID), isNull(), eq(DEVICE_ID), entryListCaptor.capture(), eq(ttlFromConfig), any(TelemetryNodeCallback.class));
+        List<TsKvEntry> entryListCaptorValue = entryListCaptor.getValue();
+        assertThat(entryListCaptorValue.size()).isEqualTo(2);
+        verifyTimeseriesToSave(entryListCaptorValue, msg, ts);
+        verify(ctxMock).tellSuccess(msg);
+        verifyNoMoreInteractions(ctxMock, telemetryServiceMock);
+    }
+
+    @ParameterizedTest
+    @MethodSource
+    void givenTtlFromConfigAndTtlFromMd_whenOnMsg_thenVerifyTtl(String ttlFromMd, long ttlFromConfig, long expectedTtl) throws TbNodeException {
+        config.setDefaultTTL(ttlFromConfig);
+        init();
+
+        when(ctxMock.getTelemetryService()).thenReturn(telemetryServiceMock);
+        when(ctxMock.getTenantId()).thenReturn(TENANT_ID);
+
+        String data = """
+                {
+                    "temp": 45,
+                    "humidity": 77
+                }
+                """;
+        var metadata = new HashMap<String, String>();
+        metadata.put("TTL", ttlFromMd);
+        TbMsg msg = TbMsg.newMsg(TbMsgType.POST_TELEMETRY_REQUEST, DEVICE_ID, new TbMsgMetaData(metadata), data);
+        node.onMsg(ctxMock, msg);
+
+        verify(telemetryServiceMock).saveAndNotify(eq(TENANT_ID), isNull(), eq(DEVICE_ID), anyList(), eq(expectedTtl), any(TelemetryNodeCallback.class));
+    }
+
+    private static Stream<Arguments> givenTtlFromConfigAndTtlFromMd_whenOnMsg_thenVerifyTtl() {
+        return Stream.of(
+                Arguments.of("5", 1L, 5L),
+                Arguments.of("", 3L, 3L),
+                Arguments.of(null, 8L, 8L)
+        );
+    }
```

Comment:
What will be if the value from metadata is 0 and tenantProfileDefaultStorageTtl is 5. Will be nice to have such a test case as well. Ideally, we should mock the tenantProfileDefaultStorageTtl in the init method to some value different from the default value in the config. This will allow us to see when the tenantProfileDefaultStorageTtl is used. Because currently we have a test: 

givenTtlFromConfigIsZeroAndUseServiceTsIsTrue_whenOnMsg_thenSaveTimeseriesUsingTenantProfileDefaultTtl

which doesn't use TTL from the tenant profile. This test works just because the value of TTL in the config equals TTL from the profile. Change the value in the config and the test will be failed.